crates/doc/src/inference.rs

```diff
@@ -1484,6 +1561,152 @@ impl Shape {
+    // NOTE: If a particular location defines `additionalProperties` as a subschema, don't
+    // add the field from `AsNode`, instead jump straight to squashing the the field into `additionalProperties`
+    pub fn widen<'n, N>(&mut self, node: &'n N, loc: Location) -> bool
+    where
+        N: AsNode,
+    {
+        match node.as_node() {
+            crate::Node::Object(fields) => {
+                // See comment in `ObjShape::widen` about when to set `is_required`
+                // on newly encountered fields. Detects whether this is the
+                // very first time this location has seen an OBJECT shape.
+                let is_first_time = !self.type_.overlaps(types::OBJECT);
+                self.type_ = self.type_ | types::OBJECT;
+
+                if let None = self.object.additional {
+                    self.object.additional = Some(Box::new(Shape {
+                        type_: types::INVALID,
+                        provenance: Provenance::Inline,
+                        ..Default::default()
+                    }))
+                }
+
+                let hint = self.object.widen::<N>(fields, loc, is_first_time);
+
+                match (hint, loc) {
+                    (true, _) => true,
+                    (false, Location::Root) => self.object.properties.len() > MAX_ROOT_FIELDS,
+                    (false, _) => self.object.properties.len() > MAX_NESTED_FIELDS,
+                }
+            }
+
+            crate::Node::Array(arr) => {
+                let mut shape = self.array.additional.take().unwrap_or(Box::new(Shape {
+                    // Start out maximally constrained
+                    type_: types::INVALID,
+                    provenance: Provenance::Inline,
+                    ..Default::default()
+                }));
+
+                // Look at each element in the observed array and widen the shape to accept it
+                let hint = arr.iter().enumerate().fold(false, |accum, (idx, node)| {
+                    accum || shape.widen(node, loc.push_item(idx))
+                });
+
+                self.array.additional = Some(shape);
+                self.type_ = self.type_ | types::ARRAY;
+
+                hint
+            }
+            crate::Node::Bool(_) => {
+                self.type_ = self.type_ | types::BOOLEAN;
+                false
+            }
+            crate::Node::Bytes(_) => {
+                self.type_ = self.type_ | types::STRING;
+
+                let partial_stringshape = StringShape {
+                    content_encoding: Some("base64".to_string()),
+                    ..Default::default()
+                };
+
+                self.string = StringShape::union(self.string.clone(), partial_stringshape);
+                false
+            }
+            crate::Node::Null => {
+                self.type_ = self.type_ | types::NULL;
+                false
+            }
+            crate::Node::Number(_num) => {
```

Comment:
you need to match over the _kind_ of number and set type_ appropriately.

See [Set::for_value](https://github.com/estuary/flow/blob/master/crates/json/src/schema/types.rs#L184) as currently used by [schema-inference](https://github.com/estuary/flow/blob/108864c33849e689da48a81f8ff7b65e02372f08/crates/schema-inference/src/inference.rs#L8) (and please double-check there isn't other handling from that implementation, which parallels this one, that we're accidentally missing).

Reply:
Done! I'm not sure what you mean by checking for "other handling from that implementation"?

Reply:
Done! I'm not sure what you mean by checking for "other handling from that implementation"? It looks like `schema-inference` just calls `Set::for_value` directly

Reply:
i meant 👀  for other things that `schema-inference` may be doing in it's own "widen"-equivalent operation, that aren't reflected here and should be.

Reply:
I took a look over at `schema-inference` and didn't see it doing anything except calling `Set::for_value`, so I think this is it 👍